openmetadata-ui/src/main/resources/ui/src/components/common/TestConnection/TestConnection.tsx

```diff
@@ -0,0 +1,340 @@
+  shouldTestConnection,
+} from 'utils/ServiceUtils';
+import {
+  showErrorToast,
+  showInfoToast,
+  showSuccessToast,
+} from 'utils/ToastUtils';
+
+import { ReactComponent as FailIcon } from 'assets/svg/fail-badge.svg';
+import { ReactComponent as SuccessIcon } from 'assets/svg/success-badge.svg';
+import { Transi18next } from 'utils/CommonUtils';
+import { TestConnectionProps, TestStatus } from './TestConnection.interface';
+import TestConnectionModal from './TestConnectionModal/TestConnectionModal';
+
+import './test-connection.style.less';
+
+// 2 minutes
+const FETCHING_EXPIRY_TIME = 2 * 60 * 1000;
+const FETCH_INTERVAL = 2000;
+const WORKFLOW_COMPLETE_STATUS = [
+  WorkflowStatus.Failed,
+  WorkflowStatus.Successful,
+];
+
+const TestConnection: FC<TestConnectionProps> = ({
+  isTestingDisabled,
+  formData,
+  serviceCategory,
+  connectionType,
+  serviceName,
+  displayButtonOnly = false,
+}) => {
+  const { t } = useTranslation();
+
+  // local state
+  const [isTestingConnection, setIsTestingConnection] =
+    useState<boolean>(false);
+  const [dialogOpen, setDialogOpen] = useState<boolean>(false);
+
+  const [message, setMessage] = useState<string>(
+    t('message.test-your-connection-before-creating-service')
+  );
+
+  const [testConnectionStep, setTestConnectionStep] = useState<
+    TestConnectionStep[]
+  >([]);
+
+  const [testConnectionStepResult, setTestConnectionStepResult] = useState<
+    TestConnectionStepResult[]
+  >([]);
+
+  const [currentWorkflow, setCurrentWorkflow] = useState<Workflow>();
+  const [testStatus, setTestStatus] = useState<TestStatus>();
+
+  /**
+   * Current workflow reference
+   */
+  const currentWorkflowRef = useRef(currentWorkflow);
+
+  // derived variables
+  const updatedFormData = useMemo(() => {
+    return formatFormDataForSubmit(formData);
+  }, [formData]);
+
+  const serviceType = useMemo(() => {
+    return getTestConnectionType(serviceCategory);
+  }, [serviceCategory]);
+
+  const allowTestConn = useMemo(() => {
+    return shouldTestConnection(connectionType);
+  }, [connectionType]);
+
+  const isTestConnectionDisabled =
+    isTestingConnection || isTestingDisabled || !allowTestConn;
+
+  // data fetch handlers
+
+  const fetchConnectionDefinition = async () => {
+    try {
+      const response = await getTestConnectionDefinitionByName(
+        toLower(connectionType)
+      );
+
+      setTestConnectionStep(response.steps);
+    } catch (error) {
+      // we will not throw error for this API
+    }
+  };
+
+  const getWorkflowData = async (workflowId: string) => {
+    try {
+      const response = await getWorkflowById(workflowId);
+      const testConnectionStepResult = response.response?.steps ?? [];
+
+      setTestConnectionStepResult(testConnectionStepResult);
+
+      setCurrentWorkflow(response);
+
+      return response;
+    } catch (error) {
+      throw error as AxiosError;
+    }
+  };
+
+  // handlers
+  const handleTestConnection = async () => {
```

Comment:
Seems difficult method to me, Is there a way to optimise it to make it more readable?

Reply:
Sure @chirag-madlani , I will try to optimize it in my next improvement PR.